pkgs/pkgs-lib/formats.nix

```diff
@@ -274,6 +274,85 @@ rec {
 
   };
 
+  # Format for defining configuration of some PHP services, that use "config.php" approach.
+  php = {finalVariable}: let
+    phpLib = {
+      phpFile = x: "<?php\n" + x;
+      buildPhpExpression = x:
+        let
+          type = builtins.typeOf x;
+          value =
+            if type == "int"
+            then "${toString x}"
+            else if type == "bool"
+            then
+              if x
+              then "true"
+              else "false"
+            else if type == "string"
+            then "\"${x}\""
+            else if type == "list"
+            then
+              "array("
+              + lib.pipe x [
+                (map phpLib.buildPhpExpression)
+                (lib.concatStringsSep ", ")
+              ]
+              + ")"
+            else if type == "set"
+            then
+              if builtins.hasAttr "_secret" x
+              then "eval(file_get_contents(\"${x._secret}\"))"
+              else if builtins.hasAttr "" x && builtins.typeOf x."" == "list"
+              then
+                "array("
+                + lib.pipe x."" [
+                  (map phpLib.buildPhpExpression)
+                  (lib.concatStringsSep ", ")
+                ]
+                + ", "
+                + lib.pipe x [
+                  (lib.filterAttrs (k: v: k != ""))
+                  (lib.mapAttrsToList (k: v: "${phpLib.buildPhpExpression k} => ${phpLib.buildPhpExpression v}"))
+                  (lib.concatStringsSep ", ")
+                ]
+                + ")"
+              else
+                "array("
+                + lib.pipe x [
+                  (lib.mapAttrsToList (k: v: "${phpLib.buildPhpExpression k} => ${phpLib.buildPhpExpression v}"))
+                  (lib.concatStringsSep ", ")
+                ]
+                + ")"
+            else throw "incompatible php value: ${toString x}";
+        in
+        value;
+    };
+
+    type = with lib.types;
+      nullOr (oneOf [
+        bool
+        int
+        float
+        str
+        (attrsOf type)
+        (listOf type)
+      ])
+      // {
+        description = "PHP value";
+      };
+  in {
+
+    inherit type;
+    lib = phpLib;
+
+    generate = name: value: pkgs.writeTextFile {
+      name = "config.php";
+      text = phpLib.phpFile "${finalVariable} = ${phpLib.buildPhpExpression value };";
+    };
```

Comment:
Taking a step back, can't we just encode to a JSON file and generate some static PHP to decode that? I don't write PHP, but something like

```nix
text = ''
  <?php 
  $json_data = json_decode(file_get_contents('${json.generate "${name}-json" value}'),true); 
  ?>
''
```

This way we can also reuse `pkgs.format.json`'s `type` and semantics.

Reply:
We can't represent array that can be both an ordinary list and an key-value map with JSON, that's why I made this code.

Reply:
What's the problem with converting JSON objects to associative arrays and JSON lists to indexed arrays?

Reply:
There's no problem if it's only **assoctiative array** or **indexed array**. But PHP allows array to be both at the same time.

Currently it's looks like this:
```nix
default = {
   "" = [ "exampleauth:UserPass" ];
   "user:password" = {
       uid = [ "user" ];
       cn = [ "user" ];
       mail = [ "user@nixos.org" ];
   };
};
```

Reply:
I created helper function `format.lib.mkMixedArray` that implements that without empty string attribute hack.